roomserver/internal/input/input_events.go

```diff
@@ -228,10 +302,115 @@ func (r *Inputer) processRoomEvent(
 	return nil
 }
 
+func (r *Inputer) checkForMissingAuthEvents(
+	ctx context.Context,
+	logger *logrus.Entry,
+	event *gomatrixserverlib.HeaderedEvent,
+	auth *gomatrixserverlib.AuthEvents,
+	known map[string]*types.Event,
+	servers []gomatrixserverlib.ServerName,
+) error {
+	authEventIDs := event.AuthEventIDs()
+	if len(authEventIDs) == 0 {
+		return nil
+	}
+
+	unknown := map[string]struct{}{}
+
+	authEvents, err := r.DB.EventsFromIDs(ctx, authEventIDs)
+	if err != nil {
+		return fmt.Errorf("r.DB.EventsFromIDs: %w", err)
+	}
+	for _, event := range authEvents {
+		if event.Event != nil {
+			ev := event // don't take the address of the iterated value
+			known[event.EventID()] = &ev
+			if err = auth.AddEvent(event.Event); err != nil {
+				return fmt.Errorf("auth.AddEvent: %w", err)
+			}
+		} else {
+			unknown[event.EventID()] = struct{}{}
+		}
+	}
+
+	if len(unknown) > 0 {
+		var res gomatrixserverlib.RespEventAuth
+		var found bool
+		for _, serverName := range servers {
+			res, err = r.FSAPI.GetEventAuth(ctx, serverName, event.RoomID(), event.EventID())
+			if err != nil {
+				logger.WithError(err).Warnf("Failed to get event auth from federation for %q: %s", event.EventID(), err)
+				continue
+			}
+			found = true
+			break
+		}
+		if !found {
+			return fmt.Errorf("no servers provided event auth")
+		}
+
+		for _, event := range gomatrixserverlib.ReverseTopologicalOrdering(
+			res.AuthEvents,
+			gomatrixserverlib.TopologicalOrderByAuthEvents,
+		) {
+			// If we already know about this event then we don't need to store
+			// it or do anything further with it.
+			if ev, ok := known[event.EventID()]; ok && ev != nil {
```

Comment:
If an event is in `known` does that imply they are verified as well? Might be worth making that crystal clear here so people know _not_ to add events to this map if they haven't been checked.